SQLQuery_Week09_Homework.sql

```diff
@@ -0,0 +1,414 @@
+
+GO
+CREATE VIEW BookIdAndNumberOfCopiesByAuthorId
+AS
+    SELECT b.Title, s.LibraryId, s.NumberOfCopies, b.AuthorId 
+    FROM Books_Sales bs
+     INNER JOIN Books b ON b.Id = bs.BookId
+     INNER JOIN Sales s ON s.Id = bs.SalesId;
+GO
+
+GO
+CREATE FUNCTION dbo.GetNumberOfCopiesOfBooksByAuthorsEmail(
+    @customerAuthorsEmail VARCHAR(100)
+)
+RETURNS INT
+AS
+BEGIN
+    DECLARE @numberOfCopies INT;
+    SET @numberOfCopies = (SELECT SUM(bc.NumberOfCopies)
+                           FROM BookIdAndNumberOfCopiesByAuthorId bc
+                                INNER JOIN Authors a ON a.Id = bc.AuthorId
+                            GROUP BY a.Email
+                            HAVING a.Email = @customerAuthorsEmail);
+    RETURN @numberOfCopies;
+END;
+GO
+
+SELECT dbo.GetNumberOfCopiesOfBooksByAuthorsEmail('mp@email.com') as 'Number of copies';
+SELECT dbo.GetNumberOfCopiesOfBooksByAuthorsEmail('ms@email.com') as 'Number of copies';
+SELECT dbo.GetNumberOfCopiesOfBooksByAuthorsEmail('cz@email.com') as 'Number of copies';
+
+
+-- 9. Create a trigger so that when a new librarian is added to the system with no library assigned (NULL on LibraryId), he / she is automatically assigned to a library.
+
+GO
+CREATE VIEW LibrariesAndLibrarians
+AS
+    SELECT LL.Id 'Library ID', LL.Name 'Library Name', L.Id 'Librarian ID', L.Name 'Librarian Name'
+    FROM Sales s
+        INNER JOIN Librarians L ON s.LibrarianId = L.Id
+        INNER JOIN Libraries LL ON s.LibraryId = LL.Id;
+GO
+
+SELECT * FROM LibrariesAndLibrarians
+
+GO
+CREATE TRIGGER AutoAssignedLibraryToNewLibrarianWithoutLibrary
+ON Librarians
+AFTER INSERT
+AS
+BEGIN
+    INSERT INTO Sales (NumberOfCopies, LibraryId, LibrarianId)
+    VALUES (0, 1, (SELECT MAX(Id) FROM Librarians))
+END
+GO
+
+DROP TRIGGER AutoAssignedLibraryToNewLibrarianWithoutLibrary
+
+DBCC CHECKIDENT ('Librarians', RESEED, 1)
+
+SET IDENTITY_INSERT Librarians ON;
+INSERT INTO Librarians(Id, Name, HireDate, IsOnHoliday)
+VALUES (8,'Catalin Stefanescu','2022-12-12',0);
+SET IDENTITY_INSERT Librarians OFF;
+
+SELECT * FROM Librarians
+SELECT * FROM Sales
+SELECT * FROM LibrariesAndLibrarians
+SELECT * FROM TypesOfBooks
+
+-- 10. Create a transaction that should have at least three commands inside it. Based on the logic you think of use at least one savepoint and at least one rollback operation.
+
+BEGIN TRANSACTION
+UPDATE TypesOfBooks SET Type = CONCAT('Type: ', Type);
+SAVE TRANSACTION InsertWordTypeOnTypesOfBooks;
+UPDATE TypesOfBooks SET Type = CONCAT('Genre: ', Type);
+ROLLBACK TRANSACTION InsertWordTypeOnTypesOfBooks;
+
+BEGIN TRANSACTION AddDefaultOneValueCopiesBooks
+UPDATE Sales SET NumberOfCopies = NumberOfCopies + 1;
+
+BEGIN TRANSACTION DefaulValuesCopiesBooks -- ??? REDUNDANT ????
+UPDATE Sales SET NumberOfCopies = NumberOfCopies -1;
+SAVE TRANSACTION DefaultCopiesBooks;
+UPDATE Sales SET NumberOfCopies = NumberOfCopies + 100;
+ROLLBACK TRANSACTION DefaultCopiesBooks;
```

Comment:
I can't add a comment on the .png file so I'm doing it here: it's a really good practice that you built the database diagram. I'm sure it helped you understand easier its structure and solve the requirements.